File: app/src/main/java/com/arcao/geocaching4locus/settings/fragment/DownloadingPreferenceFragment.java
package com.arcao.geocaching4locus.settings.fragment;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.CheckBoxPreference;

import com.arcao.geocaching4locus.App;
import com.arcao.geocaching4locus.R;
import com.arcao.geocaching4locus.base.fragment.AbstractPreferenceFragment;
import com.arcao.geocaching4locus.settings.widget.ListPreference;
import com.arcao.geocaching4locus.settings.widget.SliderPreference;

public class DownloadingPreferenceFragment extends AbstractPreferenceFragment {
    private boolean premiumMember;

    @Override
    public void onCreate(Bundle paramBundle) {
        super.onCreate(paramBundle);

        // Load the preferences from an XML resource
        addPreferencesFromResource(R.xml.preference_category_downloading);

        premiumMember = App.get(getActivity()).getAccountManager().isPremium();
    }

    @Override
    protected void preparePreference() {
        final CheckBoxPreference simpleCacheDataPreference = findPreference(DOWNLOADING_SIMPLE_CACHE_DATA, CheckBoxPreference.class);
        final ListPreference fullCacheDataOnShowPreference = findPreference(DOWNLOADING_FULL_CACHE_DATE_ON_SHOW, ListPreference.class);
        final SliderPreference downloadingCountOfLogsPreference = findPreference(DOWNLOADING_COUNT_OF_LOGS, SliderPreference.class);
        final ListPreference countOfCachesStepPreference = findPreference(DOWNLOADING_COUNT_OF_CACHES_STEP, ListPreference.class);
        final CheckBoxPreference disableDnfNmNaCachesPreference = findPreference(DOWNLOADING_DISABLE_DNF_NM_NA_CACHES, CheckBoxPreference.class);
        final SliderPreference disableDnfNmNaCachesLogsCountPreference = findPreference(DOWNLOADING_DISABLE_DNF_NM_NA_CACHES_LOGS_COUNT, SliderPreference.class);

        simpleCacheDataPreference.setEnabled(premiumMember);
        disableDnfNmNaCachesPreference.setEnabled(premiumMember);

        simpleCacheDataPreference.setOnPreferenceChangeListener((preference, newValue) -> {
            fullCacheDataOnShowPreference.setEnabled((Boolean) newValue);
            return true;
        });

        fullCacheDataOnShowPreference.setEnabled(simpleCacheDataPreference.isChecked() && premiumMember);
        fullCacheDataOnShowPreference.setSummary(preparePreferenceSummary(fullCacheDataOnShowPreference.getEntry(), R.string.pref_download_on_show_summary));

        downloadingCountOfLogsPreference.setEnabled(premiumMember);
        downloadingCountOfLogsPreference.setSummary(preparePreferenceSummary(String.valueOf(downloadingCountOfLogsPreference.getProgress()),
                R.string.pref_logs_count_summary));

        countOfCachesStepPreference.setSummary(preparePreferenceSummary(countOfCachesStepPreference.getEntry(), R.string.pref_step_geocaching_count_summary));
        disableDnfNmNaCachesLogsCountPreference.setSummary(preparePreferenceSummary(String.valueOf(disableDnfNmNaCachesLogsCountPreference.getProgress()), 0));

        if (!premiumMember) {
            applyPremiumTitleSign(simpleCacheDataPreference);
            applyPremiumTitleSign(fullCacheDataOnShowPreference);
            applyPremiumTitleSign(downloadingCountOfLogsPreference);
            applyPremiumTitleSign(disableDnfNmNaCachesPreference);
            applyPremiumTitleSign(disableDnfNmNaCachesLogsCountPreference);
            disableDnfNmNaCachesPreference.setChecked(false);
        }
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        super.onSharedPreferenceChanged(sharedPreferences, key);

        if (key == null)
            return;

        switch (key) {
            case DOWNLOADING_COUNT_OF_LOGS:
                SliderPreference countOfLogsPreference = findPreference(key, SliderPreference.class);
                SliderPreference disableDnfNmNaCachesLogsCountPreference = findPreference(DOWNLOADING_DISABLE_DNF_NM_NA_CACHES_LOGS_COUNT, SliderPreference.class);
                CheckBoxPreference disableDnfNmNaCachesPreference = findPreference(DOWNLOADING_DISABLE_DNF_NM_NA_CACHES, CheckBoxPreference.class);

                int newIntValue = countOfLogsPreference.getProgress();
                countOfLogsPreference.setSummary(preparePreferenceSummary(String.valueOf(newIntValue), R.string.pref_logs_count_summary));

                // additional checking if enough logs will be downloaded
                if (disableDnfNmNaCachesPreference.isChecked() && disableDnfNmNaCachesLogsCountPreference.getProgress() > newIntValue) {
                    if (newIntValue > 1) {
                        disableDnfNmNaCachesLogsCountPreference.setProgress(newIntValue);
                    } else {
                        disableDnfNmNaCachesPreference.setChecked(false);
                        disableDnfNmNaCachesLogsCountPreference.setProgress(1);
                    }
                }
                break;

            case DOWNLOADING_COUNT_OF_CACHES_STEP:
                final ListPreference countOfCachesStepPreference = findPreference(key, ListPreference.class);
                countOfCachesStepPreference.setSummary(preparePreferenceSummary(countOfCachesStepPreference.getEntry(), R.string.pref_step_geocaching_count_summary));
                break;

            case DOWNLOADING_FULL_CACHE_DATE_ON_SHOW:
                final ListPreference fullCacheDataOnShowPreference = findPreference(key, ListPreference.class);
                fullCacheDataOnShowPreference.setSummary(preparePreferenceSummary(fullCacheDataOnShowPreference.getEntry(), R.string.pref_download_on_show_summary));
                break;

            case DOWNLOADING_DISABLE_DNF_NM_NA_CACHES_LOGS_COUNT:
                disableDnfNmNaCachesLogsCountPreference = findPreference(key, SliderPreference.class);
                countOfLogsPreference = findPreference(DOWNLOADING_COUNT_OF_LOGS, SliderPreference.class);
                disableDnfNmNaCachesPreference = findPreference(DOWNLOADING_DISABLE_DNF_NM_NA_CACHES, CheckBoxPreference.class);

                newIntValue = disableDnfNmNaCachesLogsCountPreference.getProgress();
                disableDnfNmNaCachesLogsCountPreference.setSummary(preparePreferenceSummary(String.valueOf(newIntValue), 0));

                // additional checking if enough logs will be downloaded
                if (disableDnfNmNaCachesPreference.isChecked() && newIntValue > countOfLogsPreference.getProgress()) {
                    countOfLogsPreference.setProgress(newIntValue);
                }
                break;

            case DOWNLOADING_DISABLE_DNF_NM_NA_CACHES:
                disableDnfNmNaCachesPreference = findPreference(key, CheckBoxPreference.class);
                disableDnfNmNaCachesLogsCountPreference = findPreference(DOWNLOADING_DISABLE_DNF_NM_NA_CACHES_LOGS_COUNT, SliderPreference.class);
                countOfLogsPreference = findPreference(DOWNLOADING_COUNT_OF_LOGS, SliderPreference.class);

                // additional checking if enough logs will be downloaded
                if (disableDnfNmNaCachesPreference.isChecked() && disableDnfNmNaCachesLogsCountPreference.getProgress() > countOfLogsPreference.getProgress()) {
                    countOfLogsPreference.setProgress(disableDnfNmNaCachesLogsCountPreference.getProgress());
                }
                break;
        }
    }
}

File: app/src/main/java/com/arcao/geocaching4locus/settings/fragment/DownloadingPreferenceFragment.kt
package com.arcao.geocaching4locus.settings.fragment

import android.content.SharedPreferences
import androidx.preference.CheckBoxPreference
import androidx.preference.ListPreference
import com.arcao.geocaching4locus.R
import com.arcao.geocaching4locus.authentication.util.isPremium
import com.arcao.geocaching4locus.base.constants.PrefConstants.DOWNLOADING_COUNT_OF_CACHES_STEP
import com.arcao.geocaching4locus.base.constants.PrefConstants.DOWNLOADING_COUNT_OF_LOGS
import com.arcao.geocaching4locus.base.constants.PrefConstants.DOWNLOADING_DISABLE_DNF_NM_NA_CACHES
import com.arcao.geocaching4locus.base.constants.PrefConstants.DOWNLOADING_DISABLE_DNF_NM_NA_CACHES_LOGS_COUNT
import com.arcao.geocaching4locus.base.constants.PrefConstants.DOWNLOADING_FULL_CACHE_DATE_ON_SHOW
import com.arcao.geocaching4locus.base.constants.PrefConstants.DOWNLOADING_SIMPLE_CACHE_DATA
import com.arcao.geocaching4locus.base.fragment.AbstractPreferenceFragment
import com.arcao.geocaching4locus.data.account.AccountManager
import com.arcao.geocaching4locus.settings.widget.SliderPreference
import org.koin.android.ext.android.get

class DownloadingPreferenceFragment : AbstractPreferenceFragment() {

    override val preferenceResource: Int
        get() = R.xml.preference_category_downloading

    override fun preparePreference() {
        super.preparePreference()

        val premiumMember: Boolean = get<AccountManager>().isPremium

        val simpleCacheDataPreference = preference<CheckBoxPreference>(DOWNLOADING_SIMPLE_CACHE_DATA)
        val fullCacheDataOnShowPreference = preference<ListPreference>(DOWNLOADING_FULL_CACHE_DATE_ON_SHOW)
        val downloadingCountOfLogsPreference = preference<SliderPreference>(DOWNLOADING_COUNT_OF_LOGS)
        val countOfCachesStepPreference = preference<ListPreference>(DOWNLOADING_COUNT_OF_CACHES_STEP)
        val disableDnfNmNaCachesPreference = preference<CheckBoxPreference>(DOWNLOADING_DISABLE_DNF_NM_NA_CACHES)
        val disableDnfNmNaCachesLogsCountPreference =
            preference<SliderPreference>(DOWNLOADING_DISABLE_DNF_NM_NA_CACHES_LOGS_COUNT)

        simpleCacheDataPreference.apply {
            isEnabled = premiumMember
            setOnPreferenceChangeListener { _, newValue ->
                fullCacheDataOnShowPreference.isEnabled = newValue as Boolean
                true
            }
        }

        fullCacheDataOnShowPreference.apply {
            isEnabled = simpleCacheDataPreference.isChecked && premiumMember
            summary = preparePreferenceSummary(entry, R.string.pref_download_on_show_summary)
        }

        downloadingCountOfLogsPreference.apply {
            isEnabled = premiumMember
            summary = preparePreferenceSummary(progress.toString(), R.string.pref_logs_count_summary)
        }

        countOfCachesStepPreference.apply {
            summary = preparePreferenceSummary(entry, R.string.pref_step_geocaching_count_summary)
        }

        disableDnfNmNaCachesPreference.isEnabled = premiumMember

        disableDnfNmNaCachesLogsCountPreference.apply {
            summary = preparePreferenceSummary(progress.toString(), 0)
        }

        if (!premiumMember) {
            applyPremiumTitleSign(simpleCacheDataPreference)
            applyPremiumTitleSign(fullCacheDataOnShowPreference)
            applyPremiumTitleSign(downloadingCountOfLogsPreference)
            applyPremiumTitleSign(disableDnfNmNaCachesPreference)
            applyPremiumTitleSign(disableDnfNmNaCachesLogsCountPreference)

            disableDnfNmNaCachesPreference.isChecked = false
        }
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        super.onSharedPreferenceChanged(sharedPreferences, key)

        when (key) {
            DOWNLOADING_COUNT_OF_LOGS -> {
                val logsCountPreference = preference<SliderPreference>(key)
                val disableLogsCountPreference =
                    preference<SliderPreference>(DOWNLOADING_DISABLE_DNF_NM_NA_CACHES_LOGS_COUNT)
                val disablePreference = preference<CheckBoxPreference>(DOWNLOADING_DISABLE_DNF_NM_NA_CACHES)

                val logsCount = logsCountPreference.progress
                logsCountPreference.summary =
                    preparePreferenceSummary(logsCount.toString(), R.string.pref_logs_count_summary)

                // additional checking if enough logs will be downloaded
                if (disablePreference.isChecked && disableLogsCountPreference.progress > logsCount) {
                    if (logsCount > 1) {
                        disableLogsCountPreference.progress = logsCount
                    } else {
                        disablePreference.isChecked = false
                        disableLogsCountPreference.progress = 1
                    }
                }
            }

            DOWNLOADING_COUNT_OF_CACHES_STEP -> {
                preference<ListPreference>(key).apply {
                    summary = preparePreferenceSummary(entry, R.string.pref_step_geocaching_count_summary)
                }
            }

            DOWNLOADING_FULL_CACHE_DATE_ON_SHOW -> {
                preference<ListPreference>(key).apply {
                    summary = preparePreferenceSummary(entry, R.string.pref_download_on_show_summary)
                }
            }

            DOWNLOADING_DISABLE_DNF_NM_NA_CACHES_LOGS_COUNT -> {
                val disableLogsCountPreference = preference<SliderPreference>(key)
                val logsCountPreference = preference<SliderPreference>(DOWNLOADING_COUNT_OF_LOGS)
                val disablePreference = preference<CheckBoxPreference>(DOWNLOADING_DISABLE_DNF_NM_NA_CACHES)

                val logsCount = disableLogsCountPreference.progress
                disableLogsCountPreference.summary = preparePreferenceSummary(logsCount.toString(), 0)

                // additional checking if enough logs will be downloaded
                if (disablePreference.isChecked && logsCount > logsCountPreference.progress) {
                    logsCountPreference.progress = logsCount
                }
            }

            DOWNLOADING_DISABLE_DNF_NM_NA_CACHES -> {
                val disablePreference = preference<CheckBoxPreference>(key)
                val disableLogsCountPreference =
                    preference<SliderPreference>(DOWNLOADING_DISABLE_DNF_NM_NA_CACHES_LOGS_COUNT)
                val logsCountPreference = preference<SliderPreference>(DOWNLOADING_COUNT_OF_LOGS)

                // additional checking if enough logs will be downloaded
                if (disablePreference.isChecked && disableLogsCountPreference.progress > logsCountPreference.progress) {
                    logsCountPreference.progress = disableLogsCountPreference.progress
                }
            }
        }
    }
}
